### PR TITLE
Fix 263

### LINKED ISSUE #263: Implement ensemble reduction selection algorithms
- [ ] implement Ouranos K-means clustering method for reducing ensemble size (see https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0152495)

- [ ] Optionally implement KKZ method (PCIC) see https://journals.ametsoc.org/doi/full/10.1175/JCLI-D-14-00636.1

### PR DESCRIPTION
**Pull request by tlogan2000**
### Pull Request Checklist:
- [x] This PR addresses an already opened issue (for bug fixes / features) #263
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)
- [x] HISTORY.rst has been updated

### After merging to master and before closing the branch:
- [ ] bumpversion (minor / major / patch) has been called on `master` branch
- [ ] Tags have been pushed

* **What kind of change does this PR introduce?** 
Adds k-means clustering ensemble size reduction  see #263

* **Does this PR introduce a breaking change?** 
No 

* **Other information**:

This PR fixes #263 (kmeans only)

### CODE REVIEW

**Comment by tlogan2000** [comment]
@RondeauG I've implemented your clustering script into xclim.ensembles module.  It would be good to make sure that results using the new code are consistent with your original (i.e. running the clustering on your original data for which you wrote the scirpt)

**Comment by tlogan2000** [comment]
@huard just pushed another small commit

**Comment by tlogan2000** [comment]
> I think we should not generate figures as a side-effect of functions.
> The docstrings need a bit of love.
> The test suite is really nice.
> xarray allows us to specify dimensions by their names, let's use this instead of hardcoding the dimension index (n_sim, n_idx).

All valid points and should be doable.  The figures are quite useful but yes pulling them out of the main function would likely be the better option...  relatively simple i think using a dict of needed inputs etc

File: xclim/ensembles.py [hunk 1]
```diff
@@ -394,3 +404,296 @@ def _calc_perc(arr, p):
...
+    """Return a sample (selection) of ensemble members using k-means clustering. The algorithm attempts to
+    reduce the total number of ensemble members while maintaining adequate coverage of the ensemble
+    uncertainty (variance) in a N-dimensional data space (sel_criteria). K-Means clustering is carried out on the input
+    selection criteria data-array in order to group individual ensemble members into a reduced number of similar groups
+    Subsequently a single representative simulation is identified from each group.
+
+
+    Parameters
+    ----------
+    sel_criteria : xr.DataArray (NxP array)  ---  Selecton criteria data. These are the values used for clustering.
```
Comment (Zeitsperre) [inline xclim/ensembles.py]:
```suggestion
    sel_criteria : xr.DataArray
      Selecton criteria data. These are the values used for clustering where N is the number of realizations in the original 
      ensemble and P is the number of variables/indicators used in the grouping algorithm
```

File: xclim/ensembles.py [hunk 2]
```diff
@@ -394,3 +404,296 @@ def _calc_perc(arr, p):
...
+            https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0152495
+
+        {'rsq_cutoff': val} : threshold Coefficient of variation (R² value) above which to cover with the
+            grouping. val : float between 0 and 1. The R² indicates the proportion of the total variance in sel_criteria
+            that is explained by the grouping
+
+        {'n_clusters': val} : Create a user determined number of clusters. val : integer between 1 and N
+
+    Optional parameters:
+    max_clusters : integer  --  Maximum number of members to include in the output ensemble selection.
```
Comment (Zeitsperre) [inline xclim/ensembles.py]:
```suggestion
    max_clusters : int
      Maximum number of members to include in the output ensemble selection. When using 'rsq_optimize' or
      'rsq_cutoff' methods, limit the final selection to a maximum number even if method results indicate a higher value.
      Defaults to N (number ensemble members)
```

File: xclim/ensembles.py [hunk 3]
```diff
@@ -394,3 +404,296 @@ def _calc_perc(arr, p):
...
+        When using 'rsq_optimize' or 'rsq_cutoff' methods, limit the final selection to a maximum number even if method
+        results indicate a higher value. Defaults to N (number ensemble members)
+
+    variable_weights: xr.DataArray of size P  --  This weighting can be used to influence of weight of the climate
```
Comment (Zeitsperre) [inline xclim/ensembles.py]:
```suggestion
    variable_weights : xr.DataArray
      An array of size P This weighting can be used to influence of weight of the climate indices on the clustering itself.
```

File: xclim/ensembles.py [hunk 4]
```diff
@@ -394,3 +404,296 @@ def _calc_perc(arr, p):
...
+        indices on the clustering itself
+
+
+
+    model_weights: xr.DataArray of size N  --  This weighting can be used to influence which model is selected within
```
Comment (Zeitsperre) [inline xclim/ensembles.py]:
```suggestion
    model_weights : xr.DataArray 
      An array of size N. This weighting can be used to influence which model is selected within each cluster. This 
      parameter has no influence whatsoever on the clustering itself.
```

File: xclim/ensembles.py [hunk 5]
```diff
@@ -1,6 +1,16 @@
 import numpy as np
 import pandas as pd
+import scipy
 import xarray as xr
+from sklearn.cluster import KMeans
+
```
Comment (Zeitsperre) [inline xclim/ensembles.py]:
```suggestion
from typing import List
```

File: xclim/ensembles.py [hunk 6]
```diff
@@ -394,3 +404,296 @@ def _calc_perc(arr, p):
...
+        see https://scikit-learn.org/stable/modules/generated/sklearn.cluster.KMeans.html
+
+    graph: boolean --  displays a plot of R² vs. the number of clusters
+
+
+
+    Returns
+    -------
+
+    out : list -- Selected model indexes (positions)
```
Comment (Zeitsperre) [inline xclim/ensembles.py]:
```suggestion
    out : List
      Selected model indexes (positions)
```

**Comment by tlogan2000** [comment]
This branch seems to have lost the recent merges with master ...  I will have to do some cleanup here

**Comment by tlogan2000** [comment]
Closing with #280

**Pull request closed without merging**